### PR TITLE
RegressionRange model

### PR DESCRIPTION
**Pull request by ayush-1506**
Fixes : #18

### CODE REVIEW

File: bugbug/models/regressionrange.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+class RegressionRangeModel(BugModel):
+    def __init__(self, lemmatization=False):
+        BugModel.__init__(self, lemmatization)
+
+        self.sampler = RandomUnderSampler(random_state=0)
+
+        feature_extractors = [
+            bug_features.has_str(),
+            bug_features.has_regression_range(),
```
Comment (marco-c) [inline bugbug/models/regressionrange.py]: This one can't be used, as it's the label

File: bugbug/models/regressionrange.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            bug_features.severity(),
+            bug_features.keywords(
+                {"regression", "regression-wanted", "regressionwindow-wanted"}
```
Comment (marco-c) [inline bugbug/models/regressionrange.py]: I think `regression-wanted` doesn't exist, can you check?

**Comment by ayush-1506** [comment]
Current results:
```
X: (9612, 237133), y: (9612,)
Cross Validation scores:
Accuracy: f0.8384868415524132 (+/- 0.051361778275488414)
Precision: f0.9973640681389085 (+/- 0.0032007325171947974)
Recall: f0.83839125666528 (+/- 0.0527919641103338)
X_train: (246, 237133), y_train: (246,)
X_test: (962, 237133), y_test: (962,)

Top 15 Features:
1. 'data__has_str=yes' (+0.10424482822418213)
2. 'comments____code_reference_url__' (+0.07450265437364578)
3. 'comments__firefox' (-0.06015674024820328)
4. 'comments__mozilla' (-0.05655365809798241)
5. 'data__landings' (+0.05150233581662178)
6. 'comments__regression' (-0.043642375618219376)
7. 'comments__after' (-0.03617971017956734)
8. 'comments__nightly' (+0.03131578862667084)
9. 'comments__base' (-0.028338588774204254)
10. 'data__keywords-testcase=True' (-0.02698061428964138)
11. 'comments____file_reference__' (-0.02612282894551754)
12. 'data__has_crash_signature=False' (+0.025635896250605583)
13. 'comments__at' (+0.02212265133857727)
14. 'comments__thread' (-0.018686853349208832)
15. 'comments__patch' (-0.017180968075990677)
No confidence threshold - 962 classified
[[811 141]
 [  0  10]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.85      1.00      0.92      0.92      0.84       952
          0       0.07      1.00      0.85      0.12      0.92      0.86        10

avg / total       0.99      0.85      1.00      0.91      0.92      0.84       962


Confidence threshold > 0.6 - 881 classified
[[765 107]
 [  0   9]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.88      1.00      0.93      0.94      0.87       872
          0       0.08      1.00      0.88      0.14      0.94      0.89         9

avg / total       0.99      0.88      1.00      0.93      0.94      0.87       881


Confidence threshold > 0.7 - 800 classified
[[715  76]
 [  0   9]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.90      1.00      0.95      0.95      0.90       791
          0       0.11      1.00      0.90      0.19      0.95      0.91         9

avg / total       0.99      0.91      1.00      0.94      0.95      0.90       800


Confidence threshold > 0.8 - 691 classified
[[635  49]
 [  0   7]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.93      1.00      0.96      0.96      0.92       684
          0       0.12      1.00      0.93      0.22      0.96      0.94         7

avg / total       0.99      0.93      1.00      0.96      0.96      0.92       691


Confidence threshold > 0.9 - 541 classified
[[506  30]
 [  0   5]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.94      1.00      0.97      0.97      0.94       536
          0       0.14      1.00      0.94      0.25      0.97      0.95         5

avg / total       0.99      0.94      1.00      0.96      0.97      0.94       541
```

I haven't added the options similar to #426 yet. @marco-c Should I do that in this PR should I let this get merged first?

**Comment by marco-c** [comment]
> I haven't added the options similar to #426 yet. @marco-c Should I do that in this PR should I let this get merged first?

We can do that later as a separate PR.

Is the PR ready? If so, remove the "[WIP]" tag and I'll review it.

**Comment by ayush-1506** [comment]
Yeah, please review.

File: bugbug/models/regressionrange.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        self.clf = xgboost.XGBClassifier(n_jobs=16)
+        self.clf.set_params(predictor="cpu_predictor")
+
+    def get_labels(self):
+        classes = {}
+
+        for bug_data in bugzilla.get_bugs():
+            bug_id = int(bug_data["id"])
+            if (
+                "regression" in bug_data["keywords"]
```
Comment (marco-c) [inline bugbug/models/regressionrange.py]: `regression` in the keywords only means that the bug is a regression, but it doesn't tell us anything about whether a range is needed or not.

File: bugbug/models/regressionrange.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                or "regressionwindow-wanted" in bug_data["keywords"]
+            ):
+                classes[bug_id] = 1
+            elif "cf_has_regression_range" in bug_data:
+                if bug_data["cf_has_regression_range"] == "yes":
+                    classes[bug_id] = 1
+                elif bug_data["cf_has_regression_range"] == "no":
+                    classes[bug_id] = 0
+
```
Comment (marco-c) [inline bugbug/models/regressionrange.py]: As for the stepstoreproduce model, please print the stats about the two classes

File: run.py [hunk 5]
```diff
@@ -117,6 +118,10 @@
         from bugbug.models.stepstoreproduce import StepsToReproduceModel
 
         model_class = StepsToReproduceModel
+    elif args.goal == "regressionrange":
+        from bugbug.models.regressionrange import RegressionRangeModel
+
+        model_class = RegressionRangeModel
```
Comment (marco-c) [inline run.py]: This code changed a bit, you'll need to rebase on top of master and modify bugbug/models/__init__.py instead.

**Comment by ayush-1506** [comment]
New results after these changes:
```
1352 bugs have regression range
726 bugs don't have a regression range
X: (2078, 75510), y: (2078,)
Cross Validation scores:
Accuracy: f0.857200231298954 (+/- 0.0346064604588587)
Precision: f0.9072882348014417 (+/- 0.030668401283805345)
Recall: f0.8667669432918395 (+/- 0.06398063677610219)
X_train: (1338, 75510), y_train: (1338,)
X_test: (208, 75510), y_test: (208,)

Top 3 Features:
1. 'data__landings' (-0.20634451508522034)
2. 'data__has_str=yes' (+0.17601968348026276)
3. 'comments__revision' (-0.04829104617238045)
No confidence threshold - 208 classified
[[137  14]
 [  5  52]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.96      0.91      0.91      0.94      0.91      0.83       151
          0       0.79      0.91      0.91      0.85      0.91      0.83        57

avg / total       0.92      0.91      0.91      0.91      0.91      0.83       208


Confidence threshold > 0.6 - 198 classified
[[132  10]
 [  4  52]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.97      0.93      0.93      0.95      0.93      0.86       142
          0       0.84      0.93      0.93      0.88      0.93      0.86        56

avg / total       0.93      0.93      0.93      0.93      0.93      0.86       198


Confidence threshold > 0.7 - 176 classified
[[117   7]
 [  4  48]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.97      0.94      0.92      0.96      0.93      0.87       124
          0       0.87      0.92      0.94      0.90      0.93      0.87        52

avg / total       0.94      0.94      0.93      0.94      0.93      0.87       176


Confidence threshold > 0.8 - 153 classified
[[103   2]
 [  1  47]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.98      0.98      0.99      0.98      0.96       105
          0       0.96      0.98      0.98      0.97      0.98      0.96        48

avg / total       0.98      0.98      0.98      0.98      0.98      0.96       153


Confidence threshold > 0.9 - 116 classified
[[76  2]
 [ 0 38]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.97      1.00      0.99      0.99      0.97        78
          0       0.95      1.00      0.97      0.97      0.99      0.98        38

avg / total       0.98      0.98      0.99      0.98      0.99      0.97       116
```

**Pull request merged**